File: backend/app/models/billing.py
"""结算与余额模型"""

import uuid
from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from typing import Any

from sqlalchemy import (
    DECIMAL,
    JSON,
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import BaseModel


class InvoiceStatus(str, Enum):
    """结算单状态枚举

    新流程（多角色协作）：
        draft → pending_ops → pending_sales → pending_customer → customer_confirmed → completed

    旧流程（兼容历史数据）：
        draft → pending_customer → customer_confirmed → paid → completed

    任意非 completed 状态均可 → cancelled
    """

    DRAFT = "draft"  # 草稿（生成）
    PENDING_OPS = "pending_ops"  # 待运营经理确认
    PENDING_SALES = "pending_sales"  # 待销售经理确认
    PENDING_CUSTOMER = "pending_customer"  # 待客户确认（线下）
    CUSTOMER_CONFIRMED = "customer_confirmed"  # 客户已确认（自动扣款中/扣款失败可重试）
    PAID = "paid"  # 已付款（旧流程兼容）
    COMPLETED = "completed"  # 已完成（扣款成功）
    CANCELLED = "cancelled"  # 已取消


class CustomerBalance(BaseModel):
    """客户余额表"""

    __tablename__ = "customer_balances"

    customer_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("customers.id", ondelete="CASCADE"), unique=True
    )
    total_amount: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2), default=0)
    real_amount: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2), default=0)
    bonus_amount: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2), default=0)
    used_total: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2), default=0)
    used_real: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2), default=0)
    used_bonus: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2), default=0)

    # 关联
    customer = relationship("Customer", back_populates="balance")

    def __repr__(self):
        return f"<CustomerBalance {self.customer_id}>"


class RechargeRecord(BaseModel):
    """充值记录表"""

    __tablename__ = "recharge_records"

    customer_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("customers.id"), index=True)
    real_amount: Mapped[Decimal] = mapped_column(DECIMAL(12, 2), nullable=False)
    bonus_amount: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2), default=0)
    # total_amount 通过计算字段生成
    operator_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))
    payment_proof: Mapped[str | None] = mapped_column(String(255))
    remark: Mapped[str | None] = mapped_column(Text)


class PricingRule(BaseModel):
    """计费规则表"""

    __tablename__ = "pricing_rules"

    customer_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("customers.id"), index=True)
    device_type: Mapped[str | None] = mapped_column(
        String(20), nullable=True
    )  # X/N/L（包年结算时为 NULL）
    layer_type: Mapped[str | None] = mapped_column(
        String(20)
    )  # single/multi (前端可传 single_and_multi，后端拆分为两条)
    pricing_type: Mapped[str] = mapped_column(
        String(20), nullable=False
    )  # fixed(定价)/tier(阶梯)/package(包年)
    # 注意: 此字段与 Customer.price_policy (pricing/tiered/yearly) 命名不同、值域不同
    # pricing_type 描述具体计费规则的类型，price_policy 描述客户整体计费策略偏好
    unit_price: Mapped[Decimal | None] = mapped_column(DECIMAL(10, 2))
    # 多层计费类型：unified(统一) / incremental(递增)，仅 fixed + multi 时有效
    multi_floor_pricing_type: Mapped[str | None] = mapped_column(String(20), nullable=True)
    # 其他层单价（递增模式下，每多一层的额外计费价格）
    additional_floor_price: Mapped[Decimal | None] = mapped_column(DECIMAL(10, 2), nullable=True)
    tiers: Mapped[Any] = mapped_column(JSON)  # 阶梯配置
    package_type: Mapped[str | None] = mapped_column(String(20))  # A/B/C/D
    package_limits: Mapped[Any] = mapped_column(JSON)
    effective_date: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    expiry_date: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_by: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))

    # 关联
    customer = relationship("Customer", lazy="selectin")
    daily_consumptions = relationship("DailyConsumption", back_populates="pricing_rule")


class Invoice(BaseModel):
    """结算单表"""

    __tablename__ = "invoices"

    invoice_no: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    customer_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("customers.id"), index=True)
    period_start: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    period_end: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    total_amount: Mapped[Decimal] = mapped_column(DECIMAL(12, 2), nullable=False)
    discount_amount: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2), default=0)
    discount_reason: Mapped[str | None] = mapped_column(Text)
    discount_attachment: Mapped[str | None] = mapped_column(String(255))
    # final_amount = total_amount - discount_amount
    status: Mapped[str | None] = mapped_column(
        String(20), default="draft", index=True
    )  # draft/pending_ops/pending_sales/pending_customer/customer_confirmed/paid/completed/cancelled
    approver_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))
    approved_at: Mapped[str | None] = mapped_column(String(50))
    discount_applied_at: Mapped[str | None] = mapped_column(String(50))  # 减免申请时间
    customer_confirmed_at: Mapped[str | None] = mapped_column(String(50))
    customer_confirmed_by: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("users.id"), nullable=True
    )  # 客户确认操作人
    payment_proof: Mapped[str | None] = mapped_column(String(255))
    paid_at: Mapped[str | None] = mapped_column(String(50))
    completed_at: Mapped[str | None] = mapped_column(String(50))
    completed_by: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("users.id"), nullable=True
    )  # 完成结算操作人
    cancelled_at: Mapped[str | None] = mapped_column(String(50))  # 取消时间
    cancelled_by: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("users.id"), nullable=True
    )  # 取消操作人
    is_auto_generated: Mapped[bool | None] = mapped_column(Boolean, default=True)
    created_by: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))

    # 明细文件
    detail_file_path: Mapped[str | None] = mapped_column(
        String(255), nullable=True
    )  # Excel 文件路径
    detail_file_status: Mapped[str] = mapped_column(
        String(20), default="pending", nullable=False
    )  # pending/generating/completed/failed

    # 多角色协作确认追踪
    ops_confirmed_by: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("users.id"), nullable=True
    )  # 运营经理确认人
    ops_confirmed_at: Mapped[str | None] = mapped_column(
        String(50), nullable=True
    )  # 运营经理确认时间
    sales_confirmed_by: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("users.id"), nullable=True
    )  # 销售经理确认人
    sales_confirmed_at: Mapped[str | None] = mapped_column(
        String(50), nullable=True
    )  # 销售经理确认时间

    # 关联
    items = relationship("InvoiceItem", back_populates="invoice", cascade="all, delete-orphan")
    customer = relationship("Customer")

    __table_args__ = (
        Index("idx_invoice_customer_period", "customer_id", "period_start", "period_end"),
        Index("idx_invoice_status_period", "status", "period_start"),
    )


class InvoiceDiscountHistory(BaseModel):
    """结算单减免修改历史表

    每次修改减免时插入一条记录，保留完整操作历史。
    """

    __tablename__ = "invoice_discount_histories"

    invoice_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("invoices.id", ondelete="CASCADE"), index=True
    )
    discount_amount: Mapped[Decimal] = mapped_column(DECIMAL(12, 2), nullable=False)
    discount_reason: Mapped[str | None] = mapped_column(Text)
    discount_attachment: Mapped[str | None] = mapped_column(String(255))
    applied_at: Mapped[str] = mapped_column(String(50), nullable=False)  # 操作时间（精确到秒）
    applied_by: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("users.id"), nullable=True
    )  # 操作人

    __table_args__ = (Index("idx_discount_history_invoice", "invoice_id", "applied_at"),)


class InvoiceItem(BaseModel):
    """结算单明细表"""

    __tablename__ = "invoice_items"

    invoice_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("invoices.id", ondelete="CASCADE"), index=True
    )
    device_type: Mapped[str | None] = mapped_column(
        String(20), nullable=True
    )  # X/N/L（包年结算明细为 NULL）
    layer_type: Mapped[str | None] = mapped_column(String(20))  # single/multi
    quantity: Mapped[Decimal] = mapped_column(DECIMAL(10, 2), nullable=False)
    unit_price: Mapped[Decimal] = mapped_column(DECIMAL(10, 2), nullable=False)
    # subtotal = quantity * unit_price
    pricing_rule_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("pricing_rules.id"))

    # 关联
    invoice = relationship("Invoice", back_populates="items")


class ConsumptionRecord(BaseModel):
    """消费流水表"""

    __tablename__ = "consumption_records"

    customer_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("customers.id"), index=True)
    invoice_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("invoices.id"), index=True)
    amount: Mapped[Decimal] = mapped_column(DECIMAL(12, 2), nullable=False)
    bonus_used: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2), default=0)
    real_used: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2), default=0)
    balance_after: Mapped[Decimal | None] = mapped_column(DECIMAL(12, 2))


class SyncTaskLog(BaseModel):
    """同步任务日志表"""

    __tablename__ = "sync_task_logs"

    task_name: Mapped[str] = mapped_column(
        String(100), nullable=False, index=True, comment="任务名称"
    )
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, comment="任务状态"
    )  # success/partial/failed

    # 统计数据
    total_count: Mapped[int | None] = mapped_column(Integer, default=0, comment="总处理数量")
    success_count: Mapped[int | None] = mapped_column(Integer, default=0, comment="成功数量")
    failed_count: Mapped[int | None] = mapped_column(Integer, default=0, comment="失败数量")
    skipped_count: Mapped[int | None] = mapped_column(Integer, default=0, comment="跳过数量")

    # 执行时间
    executed_at: Mapped[str | None] = mapped_column(String(50), comment="执行时间")
    duration_seconds: Mapped[int | None] = mapped_column(
        Integer, nullable=True, comment="执行耗时 (秒)"
    )

    # 错误信息
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True, comment="错误信息")

    # 消费同步审计字段
    task_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("sync_tasks.id"),
        nullable=True,
        comment="关联任务ID",
    )
    operator_id: Mapped[int | None] = mapped_column(
        Integer,
        ForeignKey("users.id"),
        nullable=True,
        comment="操作人",
    )
    start_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, comment="同步开始日期"
    )
    end_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, comment="同步结束日期"
    )
    sync_mode: Mapped[str | None] = mapped_column(
        String(20),
        nullable=True,
        comment="同步模式: skip_existing/force_overwrite",
    )


class SyncTaskLogDetail(BaseModel):
    """同步任务执行明细表

    记录同步任务执行过程中的预警/错误/成功明细（含客户维度字段），
    用于同步日志页面查看执行信息并排查具体客户问题。
    """

    __tablename__ = "sync_task_log_details"

    task_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("sync_tasks.id", ondelete="CASCADE"),
        nullable=False,
        comment="任务ID",
    )
    sync_date: Mapped[date] = mapped_column(Date, nullable=False, comment="明细所属同步日期")
    level: Mapped[str] = mapped_column(
        String(10), nullable=False, comment="级别: info/warning/error"
    )
    category: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        comment="类别: order_fetch/order_match/order_save/cost_calc/data_check/system",
    )
    message: Mapped[str] = mapped_column(Text, nullable=False, comment="描述信息")

    # 客户维度字段（用于排查具体问题）
    customer_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("customers.id"), nullable=True, comment="内部客户ID"
    )
    customer_name: Mapped[str | None] = mapped_column(
        String(200), nullable=True, comment="内部客户名称"
    )
    external_customer_id: Mapped[str | None] = mapped_column(
        String(50), nullable=True, comment="外部客户ID(group_type)"
    )
    company_name: Mapped[str | None] = mapped_column(
        String(200), nullable=True, comment="外部公司名"
    )
    order_code: Mapped[str | None] = mapped_column(String(50), nullable=True, comment="订单号")

    # 聚合记录数（成功按客户+日期聚合时 > 1）
    record_count: Mapped[int] = mapped_column(
        Integer, nullable=False, default=1, comment="聚合记录数"
    )

    __table_args__ = (
        Index("idx_sync_detail_task_level", "task_id", "level"),
        Index("idx_sync_detail_task_date", "task_id", "sync_date"),
    )


class PackagePlan(BaseModel):
    """包年套餐表

    管理包年结算的套餐明细，支持"限量"和"不限量"两种模式。
    与 PricingRule.package_type 通过 package_type 字段关联。
    """

    __tablename__ = "package_plans"

    name: Mapped[str] = mapped_column(
        String(100), nullable=False, comment="套餐名称"
    )  # 如 "A 套餐"
    package_type: Mapped[str] = mapped_column(
        String(20), nullable=False, unique=True, index=True, comment="套餐类型标识"
    )  # 如 A/B/C/D
    device_type: Mapped[str | None] = mapped_column(
        String(20), nullable=True, comment="设备类型 (X/N/L)，为空表示通用"
    )
    layer_type: Mapped[str | None] = mapped_column(
        String(20), nullable=True, comment="楼层类型 (single/multi)，为空表示通用"
    )
    is_unlimited: Mapped[bool] = mapped_column(
        Boolean, default=False, nullable=False, comment="是否不限量：true=不限量，false=限量"
    )
    limit_count: Mapped[int | None] = mapped_column(
        Integer, nullable=True, comment="限量数量（is_unlimited=false 时必填，否则为空）"
    )
    base_fee: Mapped[Decimal] = mapped_column(
        DECIMAL(12, 2), nullable=False, comment="套餐基础费用（年费）"
    )
    over_limit_unit_price: Mapped[Decimal | None] = mapped_column(
        DECIMAL(10, 2),
        nullable=True,
        comment="超额单价: NULL=自动计算(base_fee/limit_count), 非NULL=自定义价格",
    )
    description: Mapped[str | None] = mapped_column(Text, nullable=True, comment="套餐描述")
    status: Mapped[str] = mapped_column(
        String(20), default="active", nullable=False, index=True, comment="状态：active/inactive"
    )


class AuditLog(BaseModel):
    """审计日志表"""

    __tablename__ = "audit_logs"

    user_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))
    action: Mapped[str] = mapped_column(String(50), nullable=False)
    module: Mapped[str] = mapped_column(String(50), nullable=False)
    record_id: Mapped[int | None] = mapped_column(Integer)
    record_type: Mapped[str | None] = mapped_column(String(50))
    changes: Mapped[Any] = mapped_column(JSON)  # {"before": {...}, "after": {...}}
    ip_address: Mapped[str | None] = mapped_column(String(45))
    operation_type: Mapped[str | None] = mapped_column(
        String(20),
        default="standard",
        comment="操作类型: standard/batch/relation/sensitive",
    )
    extra_metadata: Mapped[Any] = mapped_column(
        JSON,
        nullable=True,
        comment="扩展元数据: 批量统计、关系ID列表等",
    )

File: frontend/src/api/billing.ts
import api from './index'

// ==================== 余额管理 ====================

export interface Balance {
  id: number
  customer_id: number
  company_id?: number
  customer_name?: string
  account_type?: string
  industry_type?: string
  settlement_type?: string
  is_key_customer?: boolean
  total_amount: number
  real_amount: number
  bonus_amount: number
  used_total: number
  used_real: number
  used_bonus: number
  last_recharge_at?: string
  /** 近 30 天日均消耗（无消费记录时 null） */
  daily_avg_cost: number | null
  /** 近 30 天有消费的天数 */
  consumption_days: number
  /** 预计剩余天数（null = 无消耗或后付费） */
  days_remaining: number | null
}

export interface BalanceStats {
  total_balance: number
  total_customers: number
  this_month_count: number
  this_month_amount: number
  this_month_real_amount: number
  this_month_bonus_amount: number
  low_balance_count: number
  zero_balance_count: number
  /** days_remaining ≤ 7 的客户数 */
  burning_soon_count: number
}

export function getBalances(params?: {
  customer_id?: number
  keyword?: string
  account_type?: string
  industry?: string
  manager_id?: number
  sales_manager_id?: number
  recharge_date_from?: string
  recharge_date_to?: string
  tag_ids?: string
  is_key_customer?: string
  is_real_estate?: string
  settlement_type?: string
  balance_min?: number
  balance_max?: number
  sort_by?: string
  sort_order?: string
  page?: number
  page_size?: number
}) {
  return api.get('/billing/balances', { params })
}

export function getBalanceStats(params?: {
  keyword?: string
  industry?: string
  account_type?: string
  manager_id?: number
  sales_manager_id?: number
  is_key_customer?: boolean
  is_real_estate?: boolean
  settlement_type?: string
  tag_ids?: string
}) {
  return api.get<BalanceStats>('/billing/balance-stats', { params })
}

export function getCustomerBalance(customerId: number) {
  return api.get(`/billing/customers/${customerId}/balance`)
}

export interface RechargeParams {
  customer_id: number
  real_amount: number
  bonus_amount?: number
  payment_proof?: string
  remark?: string
}

export function recharge(data: RechargeParams) {
  return api.post('/billing/recharge', data)
}

export interface RechargeRecord {
  id: number
  customer_id: number
  real_amount: number
  bonus_amount: number
  total_amount: number
  operator_id: number
  payment_proof?: string
  remark?: string
  created_at: string
}

export function getRechargeRecords(params?: {
  customer_id?: number
  page?: number
  page_size?: number
}) {
  return api.get('/billing/recharge-records', { params })
}

// ==================== 消费记录 ====================

export interface ConsumptionRecord {
  id: number
  customer_id: number
  invoice_id?: number
  invoice_no?: string | null
  amount: number
  bonus_used: number
  real_used: number
  balance_after: number
  consumed_at: string
}

export function getConsumptionRecords(params?: {
  customer_id?: number
  page?: number
  page_size?: number
}) {
  return api.get('/billing/consumption-records', { params })
}

// ==================== 定价规则 ====================

export interface PricingRule {
  id: number
  customer_id?: number
  customer_name?: string
  device_type?: string
  layer_type?: string
  pricing_type: 'fixed' | 'tiered' | 'package'
  unit_price?: number
  multi_floor_pricing_type?: 'unified' | 'incremental'
  additional_floor_price?: number
  tiers?: Array<{ min: number; max: number | null; price: number }> | Record<string, unknown>
  package_type?: string
  package_limits?: Record<string, unknown>
  effective_date?: string
  expiry_date?: string | null
}

export function getPricingRules(params?: {
  customer_id?: number
  keyword?: string
  device_type?: string
  layer_type?: string
  pricing_type?: string
  page?: number
  page_size?: number
}) {
  return api.get('/billing/pricing-rules', { params })
}

export function createPricingRule(data: Partial<PricingRule>) {
  return api.post('/billing/pricing-rules', data)
}

export function updatePricingRule(id: number, data: Partial<PricingRule>) {
  return api.put(`/billing/pricing-rules/${id}`, data)
}

export function deletePricingRule(id: number) {
  return api.delete(`/billing/pricing-rules/${id}`)
}

// ==================== 定价规则冲突检查 ====================

export interface ConflictCheckParams {
  customer_id: number
  pricing_type: 'fixed' | 'tiered' | 'package'
  device_type?: string
  layer_type?: string
  effective_date: string
  expiry_date?: string
  exclude_id?: number
}

export interface ConflictRule {
  id: number
  pricing_type: string
  effective_date: string | null
  expiry_date: string | null
}

export interface ConflictCheckResult {
  has_conflict: boolean
  conflicting_rules: ConflictRule[]
}

export function checkPricingRuleConflict(params: ConflictCheckParams) {
  return api.get<ConflictCheckResult>('/billing/pricing-rules/check-conflict', { params })
}

// ==================== 结算单管理 ====================

export interface InvoiceItem {
  id?: number
  device_type?: string
  layer_type?: string
  quantity: number
  unit_price: number
  subtotal?: number
  additional_floor_price?: number
  multi_floor_pricing_type?: 'unified' | 'incremental'
  order_count?: number
  pricing_rule_id?: number
}

export interface Invoice {
  id: number
  invoice_no: string
  customer_id: number
  customer_name?: string
  /** 客户指定的运营经理 ID（用于前端判断确认按钮是否可点击） */
  customer_manager_id?: number | null
  /** 客户指定的销售经理 ID（用于前端判断确认按钮是否可点击） */
  customer_sales_manager_id?: number | null
  period_start: string
  period_end: string
  total_amount: number
  discount_amount?: number
  discount_reason?: string
  discount_attachment?: string
  final_amount: number
  status:
    | 'draft'
    | 'pending_ops'
    | 'pending_sales'
    | 'pending_customer'
    | 'customer_confirmed'
    | 'paid'
    | 'completed'
    | 'cancelled'
  is_auto_generated: boolean
  items?: InvoiceItem[]
  approver_id?: number
  approver_name?: string | null
  approved_at?: string
  discount_applied_at?: string
  ops_confirmed_by?: number | null
  ops_confirmed_name?: string | null
  ops_confirmed_at?: string
  sales_confirmed_by?: number | null
  sales_confirmed_name?: string | null
  sales_confirmed_at?: string
  customer_confirmed_at?: string
  customer_confirmed_by?: number | null
  customer_confirmed_name?: string | null
  paid_at?: string
  completed_at?: string
  completed_by?: number | null
  completed_name?: string | null
  cancelled_at?: string
  cancelled_by?: number | null
  cancelled_name?: string | null
  created_by?: number | null
  created_by_name?: string | null
  created_at: string
  /** 明细文件生成状态 */
  detail_file_path?: string | null
  detail_file_status?: string
  /** 减免修改历史（按时间倒序） */
  discount_history?: DiscountHistory[]
}

/** 减免修改历史记录 */
export interface DiscountHistory {
  id: number
  discount_amount: number
  discount_reason?: string
  discount_attachment?: string
  applied_at: string
  applied_by?: number | null
  applied_by_name?: string | null
}

/** 明细文件生成状态类型 */
export type DetailFileStatus = 'pending' | 'generating' | 'completed' | 'failed'

export function getInvoices(params?: {
  customer_id?: number
  keyword?: string
  status?: string
  page?: number
  page_size?: number
}) {
  return api.get('/billing/invoices', { params })
}

export function getInvoice(id: number) {
  return api.get(`/billing/invoices/${id}`)
}

/** 批量查询结算单文件状态（轻量轮询接口） */
export function getInvoiceFileStatus(ids: number[]) {
  return api.get('/billing/invoices/file-status', { params: { ids: ids.join(',') } })
}

export interface GenerateInvoiceParams {
  customer_id: number
  period_start: string
  period_end: string
  items: InvoiceItem[]
}

export function generateInvoice(data: GenerateInvoiceParams) {
  return api.post('/billing/invoices/generate', data)
}

export interface CalculateInvoiceItemsParams {
  customer_id: number
  period_start: string
  period_end: string
}

export function calculateInvoiceItems(data: CalculateInvoiceItemsParams) {
  return api.post('/billing/invoices/calculate-items', data)
}

// ==================== 批量生成结算单 ====================

export interface BatchGenerateParams {
  pricing_type?: 'fixed' | 'tiered' | 'package'
  industry_type_ids?: number[]
  scale_levels?: string[]
  consume_levels?: string[]
  is_real_estate?: boolean
}

export interface PreviewBatchCustomer {
  id: number
  name: string
  manager_id: number | null
  sales_manager_id: number | null
  has_manager: boolean
  has_sales_manager: boolean
}

export interface PreviewBatchResult {
  total: number
  customers: PreviewBatchCustomer[]
}

export function previewBatchCustomers(data: BatchGenerateParams) {
  return api.post<PreviewBatchResult>('/billing/invoices/preview-batch', data)
}

export interface GenerateBatchParams extends BatchGenerateParams {
  period_start: string
  period_end: string
}

export interface BatchGenerateResult {
  success_count: number
  generated: Array<{
    customer_id: number
    name: string
    invoice_id: number
    invoice_no: string
    total_amount: number
  }>
  skipped: Array<{
    customer_id: number
    name: string
    reason: string
  }>
}

export function generateInvoicesBatch(data: GenerateBatchParams) {
  return api.post<BatchGenerateResult>('/billing/invoices/generate-batch', data)
}

// ==================== 多角色确认流程 ====================

export function confirmOps(invoiceId: number) {
  return api.post(`/billing/invoices/${invoiceId}/confirm-ops`)
}

export function confirmSales(invoiceId: number) {
  return api.post(`/billing/invoices/${invoiceId}/confirm-sales`)
}

export function retryDeduction(invoiceId: number) {
  return api.post(`/billing/invoices/${invoiceId}/retry-deduction`)
}

export interface ApplyDiscountParams {
  discount_amount: number
  discount_reason?: string
  discount_attachment?: string
}

export function applyDiscount(invoiceId: number, data: ApplyDiscountParams) {
  return api.put(`/billing/invoices/${invoiceId}/discount`, data)
}

/**
 * 上传减免附件文件
 * 注意：不要手动设置 Content-Type，让浏览器自动生成包含 boundary 的 multipart 头
 */
export function uploadDiscountAttachment(file: File) {
  const formData = new FormData()
  formData.append('file', file)
  return api.post('/files/upload', formData)
}

export function submitInvoice(
  invoiceId: number,
  discount?: { discount_amount?: number; discount_reason?: string; discount_attachment?: string }
) {
  return api.post(`/billing/invoices/${invoiceId}/submit`, discount || {})
}

export function downloadInvoiceDetail(id: number) {
  return api.get(`/billing/invoices/${id}/download-detail`, { responseType: 'blob' })
}

export function regenerateInvoiceDetail(id: number) {
  return api.post(`/billing/invoices/${id}/regenerate-detail`)
}

export function getInvoiceDetailLogs(params?: {
  status?: string
  customer_id?: number
  page?: number
  page_size?: number
}) {
  return api.get('/billing/invoices/detail-logs', { params })
}

export function confirmInvoice(invoiceId: number) {
  return api.post(`/billing/invoices/${invoiceId}/confirm`)
}

export function payInvoice(invoiceId: number, data: { payment_proof?: string }) {
  return api.post(`/billing/invoices/${invoiceId}/pay`, data)
}

export function completeInvoice(invoiceId: number) {
  return api.post(`/billing/invoices/${invoiceId}/complete`)
}

export function deleteInvoice(id: number) {
  return api.delete(`/billing/invoices/${id}`)
}

export function cancelInvoice(id: number) {
  return api.post(`/billing/invoices/${id}/cancel`)
}

// 获取最近结算单（简化接口）
export function getRecentInvoices(limit: number = 10) {
  return getInvoices({ page: 1, page_size: limit })
}

// ==================== 余额趋势 ====================

export interface BalanceTrendItem {
  month: string
  total_amount: number
  real_amount: number
  bonus_amount: number
}

export function getBalanceTrend(customerId: number, months: number = 6) {
  return api.get(`/billing/customers/${customerId}/balance-trend`, { params: { months } })
}

// ==================== 余额导入 ====================

export function importBalances(file: File) {
  const formData = new FormData()
  formData.append('file', file)
  return api.post('/billing/import', formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  })
}

export function downloadBalanceImportTemplate() {
  return api.get('/billing/import-template', {
    responseType: 'blob',
  })
}

// ==================== 包年套餐管理 ====================

export interface PackagePlan {
  id: number
  name: string
  package_type: string
  device_type?: string
  layer_type?: string
  is_unlimited: boolean
  limit_count?: number | null
  base_fee: number
  /** 超额单价（限量套餐超出 limit_count 后的每单位用量价格） */
  over_limit_unit_price?: number | null
  description?: string
  status: 'active' | 'inactive'
  created_at?: string
  updated_at?: string
}

export function getPackagePlans(params?: {
  keyword?: string
  status?: string
  is_unlimited?: string
  page?: number
  page_size?: number
}) {
  return api.get('/billing/package-plans', { params })
}

export function getPackagePlan(id: number) {
  return api.get(`/billing/package-plans/${id}`)
}

export function createPackagePlan(data: Partial<PackagePlan>) {
  return api.post('/billing/package-plans', data)
}

export function updatePackagePlan(id: number, data: Partial<PackagePlan>) {
  return api.put(`/billing/package-plans/${id}`, data)
}

export function deletePackagePlan(id: number) {
  return api.delete(`/billing/package-plans/${id}`)
}
